lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1723,6 +1763,177 @@ class SCFToCalyxPass : public circt::impl::SCFToCalyxBase<SCFToCalyxPass> {
+            caller.getLoc(), getGlobalOp.getType(), getGlobalOp.getName());
+      }
+      extraMemRefOperands.push_back(newOpRes);
+
+      calleeFnBody->addArgument(newOpRes.getType(), callee.getLoc());
+      BlockArgument newBodyArg = calleeFnBody->getArguments().back();
+      op->getResult(0).replaceAllUsesWith(newBodyArg);
+      op->erase();
+      extraMemRefArgs.push_back(newBodyArg);
+      extraMemRefArgTypes.push_back(newBodyArg.getType());
+    }
+
+    SmallVector<Type, 4> updatedCalleeArgTypes(
+        callee.getFunctionType().getInputs());
+    updatedCalleeArgTypes.append(extraMemRefArgTypes.begin(),
+                                 extraMemRefArgTypes.end());
+    callee.setType(FunctionType::get(callee.getContext(), updatedCalleeArgTypes,
+                                     callee.getFunctionType().getResults()));
+
+    unsigned otherArgsCount = 0;
+    SmallVector<Value, 4> calleeArgFnOperands;
+    builder.setInsertionPointToStart(callerEntryBlock);
+    for (auto arg : callee.getArguments().take_front(originalCalleeArgNum)) {
+      if (isa<MemRefType>(arg.getType())) {
+        auto memrefType = cast<MemRefType>(arg.getType());
+        auto allocOp =
+            builder.create<memref::AllocOp>(callee.getLoc(), memrefType);
+        calleeArgFnOperands.push_back(allocOp);
+      } else {
+        auto callerArg = callerEntryBlock->getArgument(otherArgsCount++);
+        calleeArgFnOperands.push_back(callerArg);
+      }
+    }
+
+    SmallVector<Value, 4> fnOperands;
+    fnOperands.append(calleeArgFnOperands.begin(), calleeArgFnOperands.end());
+    fnOperands.append(extraMemRefOperands.begin(), extraMemRefOperands.end());
+    auto calleeName =
+        SymbolRefAttr::get(builder.getContext(), callee.getSymName());
+    auto resultTypes = callee.getResultTypes();
+
+    builder.setInsertionPointToEnd(callerEntryBlock);
+    builder.create<CallOp>(caller.getLoc(), calleeName, resultTypes,
+                           fnOperands);
+  }
+
+  /// Conditionally creates an optional new top-level function; and inserts a
+  /// call from the new top-level function to the old top-level function if we
+  /// did create one
+  LogicalResult createOptNewTopLevelFn(ModuleOp moduleOp,
+                                       std::string &topLevelFunction) {
+    auto hasMemrefArguments = [](FuncOp func) {
+      return std::any_of(
+          func.getArguments().begin(), func.getArguments().end(),
+          [](BlockArgument arg) { return isa<MemRefType>(arg.getType()); });
+    };
+
+    /// We only create a new top-level function and call the original top-level
+    /// function from the new one if the original top-level has `memref` in its
+    /// argument
+    bool hasMemrefArgsInTopLevel = false;
+    for (auto funcOp : moduleOp.getOps<FuncOp>()) {
+      if (funcOp.getName() == topLevelFunction) {
+        if (hasMemrefArguments(funcOp)) {
+          hasMemrefArgsInTopLevel = true;
+        }
+      }
+    }
```

Comment:
This can be simplified (and short-circuited) using std::any_of, e.g.,

```cpp
funcOps = moduleOp.getOps<FuncOp>();
bool hasMemrefArgsInTopLevel = 
  std::any_of(
    funcOps.begin(), 
    funcOps.end(), 
    [](auto funcOp) { 
      return funcOp.getName() == topLevelFunction && hasMemrefArguments(funcOp);
    });
```
Ref: https://en.cppreference.com/w/cpp/algorithm/all_any_none_of